Implement Go-like concurrency

omerfirmak opened the pull request:
Solves #41
Solves #40 
Solves #49

AndrejMitrovic commented:
Do you think it would make sense to split up the buffered and non-buffered implementation into separate classes with a common interface? Granted this would introduce virtual calls, and you couldn't use `() (auto ref T)` then (because templates aren't virtual).

It's just that the code is a bit complex to follow when the class has multiple modes of operation.

AndrejMitrovic commented on source/geod24/concurrency.d:
```diff
@@ -1006,6 +1007,601 @@ protected:
+                If no reader is in the wait queue and the buffer is not full
+                writer puts the message in the Channel buffer and returns immediatly
+
+                If no reader is in the wait queue and the buffer is full
```
> and the buffer is full

I don't see a check for the buffer length inside of `write()`. Is this implicit?

omerfirmak replied:
Actually there can't be any readers in the queue with a non-empty buffer. That could have been worded better, I will update it.

AndrejMitrovic commented on source/geod24/concurrency.d:
```diff
@@ -1006,6 +1007,601 @@ protected:
+
+        ***********************************************************************/
+
+        bool write () (auto ref T val)
```
What if an Exception is thrown? None of the methods are marked `nothrow`, so it's possible the call site would have to be wrapped in both a try/catch and a check for the return value.

In particular can we recover from an exception being thrown?

omerfirmak replied:
There is no reason for them to throw Exceptions. I will mark them all `nothrow`.

AndrejMitrovic commented on source/geod24/concurrency.d:
```diff
@@ -1006,6 +1007,601 @@ protected:
+                writer appends itself to write wait queue and blocks
+
+            If channel is closed, it returns immediatly with a failure,
+            regardless of the operation mode
```
This needs to clarify if it returns false **only** if the channel is closed, or if there are other cases where write can return false.

omerfirmak replied:
It is the only case, added an explanation for that.

AndrejMitrovic commented on source/geod24/concurrency.d:
```diff
@@ -1006,6 +1007,601 @@ protected:
+        ***********************************************************************/
+
+        bool write () (auto ref T val)
+        {
```
How come there is no lock in here? There doesn't need to be?

omerfirmak replied:
`tryWrite` acquires the lock. It releases the lock if it successfully completes a non-blocking operation.
In case of a blocking operation; `tryWrite` returns with the mutex still locked and `enqueueEntry` unlocks when it is finished.

AndrejMitrovic replied:
Ah yes. That makes sense, thanks!

AndrejMitrovic commented on source/geod24/concurrency.d:
```diff
@@ -1006,6 +1007,601 @@ protected:
+            }
+            else // Non-blocking operation failed and sel_state is consumed
+            {
+                this.lock.unlock();
```
This is a bit strange to me because I don't see where the lock is initially acquired (it's outside this method maybe).

omerfirmak replied:
`tryRead`/`tryWrite` are always the first ones to acquire the locks.

AndrejMitrovic commented on source/geod24/concurrency.d:
```diff
@@ -1006,6 +1007,601 @@ protected:
+
+    ***********************************************************************/
+
+    class SelectState
```
I recommend marking inner classes static in order to avoid the addition of unnecessary context pointers (it makes the class more expensive to allocate).

In addition if the class is not supposed to be derived from, mark it `final`. It should help with de-virtualizing methods (all public class methods in D are virtual).

```suggestion
    final static class SelectState
```

omerfirmak replied:
I pushed a commit for this, could you check if that is idiomatic D?

AndrejMitrovic replied:
It looks good!

AndrejMitrovic commented on source/geod24/concurrency.d:
```diff
@@ -1006,6 +1007,601 @@ protected:
+
+    ***********************************************************************/
+
+    class Channel (T)
```
```suggestion
    final static class Channel (T)
```

omerfirmak commented:
> Do you think it would make sense to split up the buffered and non-buffered implementation into separate classes with a common interface? Granted this would introduce virtual calls, and you couldn't use `() (auto ref T)` then (because templates aren't virtual).
> 
> It's just that the code is a bit complex to follow when the class has multiple modes of operation.

A standalone unbuffered channel would definitely be simpler, but a standalone buffered implementation might not be any simpler than the current hybrid implementation we have. We don't have anything extra in there to make it also work unbuffered. Unbuffered channels are just buffered channels with a full buffer, they come for free when you implement buffered ones :) 

Most of the complexity comes from the select operations. I originally wanted them to be on a completely separate layer, but baking them into the channel itself made more sense later.

AndrejMitrovic commented on source/geod24/concurrency.d:
```diff
@@ -1006,6 +1007,601 @@ protected:
+
+        void selectWrite() (T* val, SelectState sel_state, uint sel_id)
+        {
+            bool success = tryWrite(*val, sel_state, sel_id);
```
```suggestion
            assert(val !is null);
            assert(sel_state !is null);
            bool success = tryWrite(*val, sel_state, sel_id);
```

Alternatively:

```
void selectWrite() (T* val, SelectState sel_state, uint sel_id)
in
{
    assert(val !is null);
    assert(sel_state !is null);
}
body
{
}
```

I think since recently you can also write `do` instead of `body`.

AndrejMitrovic commented on source/geod24/concurrency.d:
```diff
@@ -1006,6 +1007,601 @@ protected:
+            else // Non-blocking operation failed and sel_state is consumed
+            {
+                this.lock.unlock();
+                if (sel_state.id == -1)
```
Please don't do this. You are comparing an unsigned type `uint` with `-1`.

AndrejMitrovic replied:
I also don't see where it's ever initialized to `-1` / `uint.max`.

omerfirmak replied:
I wish there was a compiler warning for this, will make it an `int`. 
`id` can be set to `-1` in `dequeueEntry`

AndrejMitrovic replied:
> I wish there was a compiler warning for this

We do, too. But Walter Bright is against it. :P

AndrejMitrovic commented on source/geod24/concurrency.d:
```diff
@@ -1006,6 +1007,601 @@ protected:
+                    this.buffer = this.buffer.remove(0);
+
+                    if (write_ent)
+                    {
```
```suggestion
                    {
                        assert(write_ent.pVal !is null);
```

omerfirmak replied:
Maybe this should be done at the enqueue stage to better localize the source.

AndrejMitrovic commented on source/geod24/concurrency.d:
```diff
@@ -1006,6 +1007,601 @@ protected:
+            // if dequeueEntry returns a valid entry, it always successfully consumes the related select states.
+            // the race between 2 select calls is resolved in dequeueEntry.
+            else if (write_ent)
+            {
```
```suggestion
            {
                assert(write_ent.pVal !is null);
```

AndrejMitrovic commented on source/geod24/concurrency.d:
```diff
@@ -1006,6 +1007,601 @@ protected:
+            {
+                this.lock.unlock();
+
+                if (sel_state.id == -1)
```
Same comment about comparing unsigned type to `-1`.

AndrejMitrovic commented on source/geod24/concurrency.d:
```diff
@@ -1006,6 +1007,601 @@ protected:
+
+        private ChannelQueueEntry enqueueEntry(ref ChannelQueueEntry[] entryq, T* pVal, SelectState sel_state = null, uint sel_id = 0)
+        {
+            ChannelQueueEntry q_ent = new ChannelQueueEntry(pVal, sel_state, sel_id);
```
Why does a `ChannelQueueEntry` need to be a class? It seems like a lot of overhead. It might be better to just use a struct and avoid unnecessary GC allocations.

omerfirmak replied:
I needed it to be a heap-allocated-reference-type and went for a class.
Would this function do any allocations other than the one we do explicitly here? I assumed that only the reference of the `q_ent` gets passed around. Wouldn't a heap allocated struct still use GC?

AndrejMitrovic replied:
Ah if you need reference semantics, it's ok to keep it as a class. I didn't realize that was the use-case.

AndrejMitrovic replied:
> Wouldn't a heap allocated struct still use GC?

Yes. I wrongly assumed we can use value semantics here.

AndrejMitrovic reviewed (commented):
First wave of reviews. I'll have more coming later.

omerfirmak commented:
> First wave of reviews. I'll have more coming later.

Thank you for the review! I pushed separate commits for each comment. We can squash them once we are happy with the changes.

AndrejMitrovic commented on source/geod24/concurrency.d:
```diff
@@ -897,6 +898,78 @@ class FiberScheduler
+        {
+            this.registerToInfoFiber();
+            FiberScheduler.yield();
+            scope (exit) this.notified = false;
```
I think this is in the wrong position. It should be placed at the start of the wait function.

AndrejMitrovic replied:
Hold on, actually the function is nothrow. So you can just remove `scope (exit)` entirely.

omerfirmak replied:
Yes, I will remove it. I know the order of scope statements matter when there is multiple. They get executed in reverse. But does it matter when there is only one?

Geod24 commented on source/geod24/concurrency.d:
```diff
@@ -934,76 +1007,616 @@ protected:
+
+            Buffered mode:
+
+                If a reader is already blocked on the Channel, writer copies the
+                message to reader's buffer and wakes up the reader by yielding.
+                After execution comes back to the writer, it returns immediatly.
```
Technically this differs from the Golang semantic. In Go, writing to an unbuffered channel *always* block the caller (either really blocking, or hand off to pending reader), but writing to a buffered channel never blocks for the writer if the channel isn't full.

Example code:
```Go
package main

import (
	"fmt"
)

func receiver(ch chan string) {
	foo := <- ch
	fmt.Println("Received: ", foo)
}

func main() {
         // Create an unbuffered channel
	ch := make(chan string)
	// Create a buffered channel
        ch := make(chan string, 2)
	go receiver(ch)
	ch <- "World"
	ch <- "World"	 // Remove me if using unbuffered channel
	fmt.Println("Hello, playground")
}
```

If you run it on the playground, [you can see](https://play.golang.org/p/ZqcJvc1pCsk) that it only prints "Hello playground", never prints the "Hello" sent.

Now, the title of the issue mentioned "Go style concurrency" not because we need to emulate Go in every aspect, but because we wanted to provide a directional, pipe based API. So while we don't have to emulate every aspects of it, I wonder if this  semantic is actually relied on somewhere, and how useful it actually is.

Thinking about it, it would make sense to have a buffered channel when your writer has high throughput, and unbuffered channel when the reader does (because of the hands off semantic).

omerfirmak replied:
This is solely because the `FiberBinarySemaphore` yields the running Fiber on `notify()` calls. I tried to change that before, but the agora tests failed.

omerfirmak replied:
This is closer to Golang behavior with the new FiberScheduler in this [branch](https://github.com/omerfirmak/localrest/tree/go-style-sched).

Geod24 commented on source/geod24/concurrency.d:
```diff
@@ -897,6 +898,78 @@ class FiberScheduler
+
+        }
+
+        void wait() nothrow
```
Style nitpick: We always put space between a parenthesis and a function declaration or definition.
So this should be:
```suggestion
        void wait () nothrow
```

On the other hand, if you call it, no space:
```D
foo.wait();
```

This makes grepping / Ctrl + F for the definition super simple.

omerfirmak replied:
Noted.

omerfirmak replied:
Applied this style to functions I recently touched or wrote myself.

AndrejMitrovic commented on source/geod24/concurrency.d:
```diff
@@ -934,76 +1007,616 @@ protected:
+            if (this.buffer.length > 0)
+            {
+                // if dequeueEntry fails, we will try to consume caller_sel again.
+                if (!caller_sel || write_ent || caller_sel.tryConsume(caller_sel_id))
```
Won't this segfault if `caller_sel` is null and `write_ent` is true?

omerfirmak replied:
If `caller_sel` is null, the latter 2 expressions won't get evaluated.

AndrejMitrovic replied:
Oh right, I misread it! Personally I would make this a bit clearer:

```suggestion
                if (write_ent || (caller_sel && caller_sel.tryConsume(caller_sel_id))
```

I think that would have the same semantics?

omerfirmak replied:
We still want to execute the body of `if` when `caller_sel` is null. I think this would fail in that case, when `write_ent` is null.

AndrejMitrovic replied:
Ah yep. Carry on. :)

Geod24 commented on source/geod24/concurrency.d:
```diff
@@ -934,76 +1007,616 @@ protected:
-    protected class FiberBinarySemaphore
+    /***********************************************************************
+
+        SelectState
```
Did you read [the documentation specs](https://dlang.org/spec/ddoc.html) ?
If you want some example of how things would look like, I recommend you look at the agora docs.
In particular, [a simple module like ECC](https://bpfkorea.github.io/agora/agora/common/crypto/ECC.html) would produce [this documentation](https://bpfkorea.github.io/agora/agora/common/crypto/ECC.html).

You can see what looks readable and what doesn't, e.g. if you go to the `Scalar` struct and check the functions, you'll see which ones are well documented, which one is not, and which one should have its description split in two ;)

omerfirmak replied:
I didn't :) Will take a look now.

Geod24 commented:
I will give this a test, but I don't see any handling of `shared` vs non-`shared` data, which might wreak havoc in the design if not considered from early on.

omerfirmak commented:
> I will give this a test, but I don't see any handling of `shared` vs non-`shared` data, which might wreak havoc in the design if not considered from early on.

Please be gentle :)

omerfirmak commented:
@AndrejMitrovic @Geod24  What do you think about the `select()` implementation I just pushed? Especially about the `void*` and pointer casts. Should we have a protection against misuse?

btw, I didn't bother with doc and styling just yet.

AndrejMitrovic commented:
> What do you think about the select() implementation I just pushed?

I'm not sure. What is the benefit?

omerfirmak commented:
> > What do you think about the select() implementation I just pushed?
> 
> I'm not sure. What is the benefit?

Mainly, it lets user to programmatically select on multiple channels without hard coding `selectRead`/`selectWrite` calls. It also lets us manage the select flow and hide some implementation details. For example, `selectRead`/`selectWrite` can fail (because of a low probability race condition) but we can make sure `select()` does not by looping until we succeed. A `select()` with a timeout will need some special handling, which we can do correctly in `select()` and not rely on user to get it right every time.

Geod24 commented on source/geod24/concurrency.d:
```diff
@@ -990,150 +966,889 @@ protected:
+
+        ***********************************************************************/
+
+        ulong length () nothrow
```
```suggestion
        size_t length () nothrow
```

Geod24 commented:
Wow, great work! It's fairly hard to assess the impact at the moment, but this is definitely going in the right direction. Kudos for the care you put into documenting it, highly appreciated.

omerfirmak committed:
Port LocalRest to use new concurrency primitives

omerfirmak commented:
I have broken the code into 5 commits. We can go commit by commit and merge them all at once or have them each on a seperate PR. But it makes more sense to me to merge them all at once. @Geod24 @AndrejMitrovic

AndrejMitrovic commented on source/geod24/array.d:
```diff
@@ -0,0 +1,315 @@
+/**
```
I think this module should be part of its own commit.

AndrejMitrovic commented on source/geod24/concurrency.d:
```diff
@@ -823,6 +861,7 @@ public class InfoThread : Thread
  */
 class FiberScheduler
 {
+
```
boop

AndrejMitrovic commented on source/geod24/concurrency.d:
```diff
@@ -815,6 +820,39 @@ public class InfoThread : Thread
+
+***************************************************************************/
+
+public void thisScheduler (FiberScheduler value) nothrow
```
I guess you could mark this as a `@property` and then use it like `thisScheduler = null` for example. Because `thisScheduler(arg)` looks really strange.

omerfirmak replied:
They were initially marked as a property but I ran into a problem. Will check it again.

AndrejMitrovic replied:
I wonder if it's something to do with overloads.

Well, `@property` is a bit weird because it's got some edge-cases. You might have ran into them.

AndrejMitrovic commented on source/geod24/concurrency.d:
```diff
@@ -946,39 +985,52 @@ protected:
         }
 
-        void wait() nothrow
+        void wait () nothrow
```
In the future please make these styling changes in their own commit.

AndrejMitrovic commented on source/geod24/concurrency.d:
```diff
@@ -1057,6 +1111,8 @@ private:
             }
         }
+
+        thisScheduler(null);
```
Not in a `scope (exit)` at the beginning? I guess it doesn't matter after an exception is thrown..

omerfirmak replied:
That should be scoped regardless, just to be safe.

AndrejMitrovic commented on source/geod24/concurrency.d:
```diff
@@ -1446,3 +1502,665 @@ private:
+
+final public class Channel (T)
+{
+
```
boop

AndrejMitrovic commented on source/geod24/concurrency.d:
```diff
@@ -1446,3 +1502,665 @@ private:
+
+***********************************************************************/
+
+final public class Channel (T)
```
`T` needs `Parameter` documentation. Are there any constraints here? T can be anything?

omerfirmak replied:
Are there any examples of this? Cant find a special keyword for template documentation in https://dlang.org/spec/ddoc.html

Geod24 replied:
It's not special, it is simply a parameter for the documentation generator.

AndrejMitrovic commented on source/geod24/concurrency.d:
```diff
@@ -1446,3 +1502,665 @@ private:
+
+        Param:
+            max_size = Maximum amount of T a Channel can buffer
+                            (0 -> Unbuffered operation,
```
Unindent please so the text aligns.

AndrejMitrovic commented on source/geod24/concurrency.d:
```diff
@@ -1446,3 +1502,665 @@ private:
+        successfully consume the `caller_sel`
+
+        Param:
+            output = Field to write the message to
+            caller_sel = SelectState instace of the select call being executed
+            caller_sel_id = id of the select call being executed
```
This parameter documentation is wrong.

AndrejMitrovic commented on source/geod24/concurrency.d:
```diff
@@ -1446,3 +1502,665 @@ private:
+
+            if (write_ent)
+            {
+                this.buffer.putBack(*(write_ent.pVal));
```
`this.buffer.putBack(*write_ent.pVal)` works fine too.

AndrejMitrovic commented on source/geod24/concurrency.d:
```diff
@@ -1446,3 +1502,665 @@ private:
+        // the race between 2 select calls is resolved in dequeueEntry.
+        else if (write_ent)
+        {
+            output = *(write_ent.pVal);
```
output = *write_ent.pVal;

AndrejMitrovic commented on source/geod24/concurrency.d:
```diff
@@ -1446,3 +1502,665 @@ private:
+
+    size_t length () nothrow
+    {
+        this.lock.lock_nothrow();
```
@omerfirmak my concurrency-foo isn't the best.

Can you help me understand the purpose of using a lock when looking up a value like this? Because as soon as the function returns the lock is released, which means the state of the channel could change while the calling code still assumes the channel is in the same state.

Isn't it the same to just return the value directly?

AndrejMitrovic replied:
Also I feel like for checking lengths and booleans we should use `atomicStore` and `atomicLoad`, because locking is expensive.

omerfirmak replied:
You are right. As long as every operation (load/store) on the variables are atomic, there is no need to use a lock in these getters. We can do this easily with the closed. But since the length is managed by the ring buffer implementation, I did not want to prematurely optimize. We can definitely get rid of those once the implementation is complete, feature and correction wise.

AndrejMitrovic replied:
Good idea. Safety first.

AndrejMitrovic commented on source/geod24/concurrency.d:
```diff
@@ -1446,3 +1502,665 @@ private:
+        T* pVal;
+
+        /// Result of the blocking read/write call
+        bool success = true;
+
+        this (T* pVal) nothrow
+        {
+            this.pVal = pVal;
+            this.sem = thisScheduler().new FiberBinarySemaphore();
+        }
+
+        /***********************************************************************
+
+            Terminate a `ChannelQueueEntry` by waking up the blocked Fiber
+            and reporting the failure
+
+            This is called on all the `ChannelQueueEntry` instances still in
+            the wait queues when Channel is closed
+
+        ***********************************************************************/
+
+        void terminate () nothrow
+        {
+            this.success = false;
+            this.sem.notify();
+        }
+    }
+
+    /***********************************************************************
+
+        Create and enqueue a `ChannelQueueEntry` to the given entryq
+
+        Assumes that the Channel is locked, unlocks it before returning
+
+        Param:
+            entryq = Queue to append the new ChannelQueueEntry
+            pVal =  Pointer to the message buffer
+            sel_state = SelectState object to associate with the
+                        newly created ChannelQueueEntry
+            sel_id = id of the select call creating the new ChannelQueueEntry
```
Wrong docs, there are only 2 params.

AndrejMitrovic commented on source/geod24/concurrency.d:
```diff
@@ -1446,3 +1502,665 @@ private:
+
+    private ChannelQueueEntry dequeueEntry (ref DList!ChannelQueueEntry entryq) nothrow
+    {
+        while (!entryq.empty())
```
This should be an `if`.

omerfirmak replied:
Yep, that must have left over while manually I split the code into commits :(

AndrejMitrovic commented on source/geod24/concurrency.d:
```diff
@@ -1446,3 +1502,665 @@ private:
+            assert(reader_sum == writer_sum);
+        }
+    );
+}
```
I think you'll need to change your text editor settings. Needs trailing newline. :>

AndrejMitrovic commented on source/geod24/concurrency.d:
```diff
@@ -1503,6 +1503,152 @@ private:
     assert(x[0] == 5);
 }
 
+/***********************************************************************
+
+    A common interface for objects that can be used in `select()`
+
+***********************************************************************/
+
+public interface Selectable
+{
+    void selectWrite (void* ptr, SelectState sel_state, int sel_id);
+    void selectRead (void* ptr, SelectState sel_state, int sel_id);
```
Docs please.

AndrejMitrovic commented on source/geod24/concurrency.d:
```diff
@@ -1503,6 +1503,152 @@ private:
+    }
+}
+
+public alias SelectReturn = Tuple!(bool, "success", int, "id");
```
docs, because it's public

AndrejMitrovic commented on source/geod24/concurrency.d:
```diff
@@ -1503,6 +1503,152 @@ private:
+    /// Pointer to the select data
+    void* select_data;
+
+    this (Selectable selectable, void* select_data)
```
docs please, otherwise it might seem like you can partially construct it via `SelectEntry(selectable)`.

AndrejMitrovic commented on source/geod24/concurrency.d:
```diff
@@ -1503,6 +1503,152 @@ private:
+
+    Params:
+        read_list = List of `Channel`s to select for read operation
+
```
boop

omerfirmak replied:
With no space in between those lines the doc generator did not put each param in its own line.

AndrejMitrovic replied:
Huh? That sounds like a bug. :/

AndrejMitrovic commented on source/geod24/concurrency.d:
```diff
@@ -1503,6 +1503,152 @@ private:
+    {
+        entry.selectable.selectRead(entry.select_data, ss, sel_id++);
+    }
+
```
Does `sel_id` maybe have to be reset to 0 here before the second foreach loop?

Because if so, I would suggest to just use `foreach (sel_id, ref entry; *_list)` in both foreach loops.

omerfirmak replied:
No, `sel_id` returned is the index of the SelectEntry in the `read_list ~ write_list`. It should be documented better.

AndrejMitrovic commented on source/geod24/concurrency.d:
```diff
@@ -1503,6 +1503,152 @@ private:
+
+final private class SelectState
+{
+
```
boop

AndrejMitrovic commented on source/geod24/concurrency.d:
```diff
@@ -1503,6 +1503,152 @@ private:
+        Param:
+            id = ID of the `selectRead`/`selectWrite` call that is consuming
+                    this `SelectState`
+
```
extra blank line

AndrejMitrovic commented on source/geod24/concurrency.d:
```diff
@@ -1608,22 +1754,25 @@ final public class Channel (T)
     ***********************************************************************/
 
-    private bool tryWrite () (auto ref T val) nothrow
+    private bool tryWrite () (auto ref T val, SelectState caller_sel = null, int caller_sel_id = 0) nothrow
```
I see that some of the documentation for these parameters was actually introduced in the previous commit. This should be fixed.

AndrejMitrovic commented on source/geod24/concurrency.d:
```diff
@@ -1632,6 +1781,42 @@ final public class Channel (T)
+        create a write queue entry using the given `sel_state` and `sel_id`
+
+        Param:
+            val = Message to write to channel
+
+            sel_state = SelectState instace of the select call being executed
+
+            sel_id = id of the select call being executed
```
extra blank lines

AndrejMitrovic commented on source/geod24/concurrency.d:
```diff
@@ -1632,6 +1781,42 @@ final public class Channel (T)
+        bool success = tryWrite(*val, sel_state, sel_id);
+
+        if (!sel_state.isConsumed())
+            enqueueEntry(this.writeq, val, sel_state, sel_id);
```
is this a global function? Otherwise maybe use `this.enqueueEntry`.

AndrejMitrovic commented on source/geod24/concurrency.d:
```diff
@@ -1705,18 +1890,26 @@ final public class Channel (T)
-            if (write_ent)
+                if (write_ent)
+                {
+                    this.buffer.putBack(*(write_ent.pVal));
```
`this.buffer.putBack(*write_ent.pVal);`

AndrejMitrovic commented on source/geod24/concurrency.d:
```diff
@@ -1735,6 +1930,42 @@ final public class Channel (T)
+        create a read queue entry using the given `sel_state` and `sel_id`
+
+        Param:
+            val = Buffer to write the message to
+
+            sel_state = SelectState instace of the select call being executed
+
+            sel_id = id of the select call being executed
```
extra blank lines

AndrejMitrovic commented on source/geod24/concurrency.d:
```diff
@@ -1851,11 +2095,11 @@ final public class Channel (T)
     ***********************************************************************/
 
-    private ChannelQueueEntry enqueueEntry (ref DList!ChannelQueueEntry entryq, T* pVal) nothrow
+    private ChannelQueueEntry enqueueEntry (ref DList!ChannelQueueEntry entryq, T* pVal, SelectState sel_state = null, int sel_id = 0) nothrow
```
This line is too long now.

AndrejMitrovic commented on source/geod24/concurrency.d:
```diff
@@ -1887,13 +2131,41 @@ final public class Channel (T)
     ***********************************************************************/
 
-    private ChannelQueueEntry dequeueEntry (ref DList!ChannelQueueEntry entryq) nothrow
+    private ChannelQueueEntry dequeueEntry (ref DList!ChannelQueueEntry entryq, SelectState caller_sel = null, int caller_sel_id = 0) nothrow
```
ditto line too long

AndrejMitrovic commented on source/geod24/concurrency.d:
```diff
@@ -1030,6 +1023,7 @@ protected:
 
         void notify () nothrow
         {
+            this.freezeTimer();
```
Why do you call it `freeze` rather than `stop`? Do you ever resume that same timer? That's what it seems to imply.

omerfirmak replied:
No particular reason and they dont ever get resumed. Stop would have been more appropriate. Will change.

AndrejMitrovic commented on source/geod24/concurrency.d:
```diff
@@ -1047,15 +1041,55 @@ protected:
+
+        ***********************************************************************/
+
+        bool wasTimedOut () nothrow
```
hasTimedOut?

omerfirmak replied:
Me englando no good :)

AndrejMitrovic commented on source/geod24/LocalRest.d:
```diff
@@ -110,30 +110,6 @@ import std.traits : fullyQualifiedName, Parameters, ReturnType;
 import core.thread;
 import core.time;
 
-/// Request / Response ID
```
I wish the moving of the symbols was in another commit. It makes it very hard to see what actually changed vs what was only moved.

omerfirmak replied:
I will keep that in mind.

AndrejMitrovic reviewed (changes requested):
Most of my comments were about atomicity of commits and some missing docs.

Actually it's a bit hard to review because there's a lot of code changes. Hopefully @Geod24 can give it a review some time too.

Overall though it looks good. You've added tests for most code which is great. Thanks a lot for the work so far!

omerfirmak committed:
Make FiberScheduler aware of its work queues

This extended FiberScheduler will block the running thread
when no Fiber is ready to run.

omerfirmak commented:
Thank you for the review @AndrejMitrovic ! All of your comments should be addressed now.

Geod24 commented on source/geod24/LocalRest.d:
```diff
@@ -410,6 +331,61 @@ public final class RemoteAPI (API, alias S = VibeJSONSerializer!()) : API
 
+    alias RestChannel = C.Channel!Variant;
+
+    /// Ask the node to shut down
+    private struct ShutdownCommand (API)
+    {
+        /// Any callback to call before the Node's destructor is called
+        void function (API) callback;
+
+        /// Whether we're restarting or really shutting down
+        bool restart;
+    }
+
+    /// Data sent by the caller
+    private struct Command
+    {
+        /// Channel to be used when responding to this Command
+        RestChannel resp_chn;
+        /// Method to call
+        string method;
+        /// Serialized arguments to the method
+        SerializedData args;
+    }
```
The reason those were outside of any template was to avoid template bloat.
In this case, the first struct (`ShutdownCommand`) can be moved here, but then, you should remove the template parameter and pick the one from the context, otherwise there will always be 2 (equal) template parameter (not bloat, but useless).

IIUC, the reason you moved `Command` in this function was to have access to `RestChannel`, and by extension, `Variant`. My instinct would be to rather move `Variant` *outside* of this function.

Now if you do so, you'll need to template `Command` on the channel type. But actually, you don't! The reason to have an `id` in the command was to allow respond, since we didn't have channels. Now that we have this, the node can just respond on the channel it received on.

Geod24 replied:
Wait I see `ID` has been removed. That's odd. `RestChannel` is the "connection", `ID` is used to ensure re-entrancy if multiple queries are run in parallel.

omerfirmak replied:
With a per command response channel, you should not need an id for re-entrancy. For the exact same behaviour, connection channel should be flushed when node restarts though.

Geod24 replied:
Per command ? You mean per type or per instance ? Assuming it's per instance, it's bringing its own host of problems.

omerfirmak replied:
In the current implementation, each node has a channel that it listens for commands and each command comes with a channel that the node should respond on. So it is per command instance, for now.

omerfirmak replied:
> My instinct would be to rather move Variant outside of this function.

If you move the Variant outside of the class, wouldn't you loose the API information? This, the structs being moved, was all because of the ShutdownCommand being templated.

Geod24 replied:
> If you move the Variant outside of the class, wouldn't you loose the API information? This, the structs being moved, was all because of the ShutdownCommand being templated.

Well you can do `Variant(API)` and have an alias here.

Geod24 commented on source/geod24/LocalRest.d:
```diff
@@ -440,8 +416,12 @@ public final class RemoteAPI (API, alias S = VibeJSONSerializer!()) : API
     {
-        auto childTid = C.spawn(&spawned!(Impl), file, line, args);
-        return new RemoteAPI(childTid, timeout);
+        auto chn = new RestChannel(16);
+        new Thread(
+        {
+            spawned!(Impl)(chn, file, line, args);
+        }).start();
+        return new RemoteAPI(chn, timeout);
```
Why is the channel buffered ?

omerfirmak replied:
To allow multiple requests to be queued. The requester would wait on the response channel. I am not sure if that would have any performance benefits though. This probably will have to be unbuffered for the suggestion you made below to work.

Geod24 replied:
Unbuffered also means the sender will not block, won't that create issues ?

omerfirmak replied:
Do you mean buffered? Currently it wont. the `read` on the response Channel will block the sender.

Geod24 commented on source/geod24/LocalRest.d:
```diff
@@ -532,9 +513,7 @@ public final class RemoteAPI (API, alias S = VibeJSONSerializer!()) : API
-    private static void spawned (Implementation) (
-        C.Tid self, string file, int line, CtorParams!Implementation cargs)
-        nothrow
+    private static void spawned (Implementation) (RestChannel chn, string file, int line, CtorParams!Implementation cargs) nothrow
```
Indentation. Also need to change the `Params` documentation. Note that I tried quite hard to keep a parallel with the way a TCP (/HTTP) server would work (listen & co) to make it easier to reason about.

Geod24 commented on source/geod24/LocalRest.d:
```diff
@@ -410,6 +331,61 @@ public final class RemoteAPI (API, alias S = VibeJSONSerializer!()) : API
 {
     static assert (!serializerInvalidReason!(S).length, serializerInvalidReason!S);
 
+    alias RestChannel = C.Channel!Variant;
```
I highly regret picking the name "localrest" for this repository... When I started, I just thought "Hey I'm going to mock `vibe.web.rest`"... After almost 2 years, it is very clear that it's not just mocking a REST generator, but a full network (it supports custom serialization, has a network registry, support simultaneous connections, etc...). Hence a more apt name would be "Connection".

Geod24 commented on source/geod24/LocalRest.d:
```diff
@@ -551,22 +530,6 @@ public final class RemoteAPI (API, alias S = VibeJSONSerializer!()) : API
             }
         }
 
-        // very simple & limited variant, to keep it performant.
-        // should be replaced by a real Variant later
```
I'm glad we never used the real `Variant` ;)

Geod24 commented on source/geod24/LocalRest.d:
```diff
@@ -410,6 +331,61 @@ public final class RemoteAPI (API, alias S = VibeJSONSerializer!()) : API
+        /// Type of a request
+        enum Type
+        {
+            command,
+            response,
+            filter,
+            time_command,
+            shutdown_command
```
Two things:
1) In our style guide, `enum` names are Snake (e.g. `Command`). Not a problem since they're namespaced.
2) I tend to use a trailing comma at the end, just to make it more change-friendly (D allows trailing commas pretty much anywhere and it's awesome).

omerfirmak replied:
Is the style guide posted somewhere?

AndrejMitrovic replied:
Part of it is here: https://github.com/bpfkorea/agora/blob/v0.x.x/doc/Style.md

There may be some missing notes though.

Geod24 commented on source/geod24/LocalRest.d:
```diff
@@ -410,6 +331,61 @@ public final class RemoteAPI (API, alias S = VibeJSONSerializer!()) : API
+            ShutdownCommand!API shutdown;
+        }
+
+        ubyte tag;
```
If you define an `enum`, why is this not typed ?

omerfirmak replied:
It must have carried over from the old code. Will fix.

Geod24 commented on source/geod24/LocalRest.d:
```diff
@@ -410,6 +331,61 @@ public final class RemoteAPI (API, alias S = VibeJSONSerializer!()) : API
+        ubyte tag;
+
+        /// Type of a request
+        enum Type
```
`enum Type : ubyte`

Geod24 commented on source/geod24/LocalRest.d:
```diff
@@ -595,61 +558,51 @@ public final class RemoteAPI (API, alias S = VibeJSONSerializer!()) : API
+                                FilterAPI filter_api = msg.filter;
+                                control.filter = filter_api;
+                                break;
+                            default:
+                                assert(0);
```
Note: `enum` are absolutely effing bs because you need to sanitize them EVERY. EFFING. TIME.
So even if your tag should only take values `[0 .. 4]` if correctly typed, you can actually receive `[0 .. 255]`.
`final switch` allows you to omit the `default` if you handle all values, but obviously that's just bad.
In this case, throwing an `Error` with the enum value would be good (there's a catch at the top level that will print an error message).

Geod24 replied:
Rule of thumb: Assume input to your function is provided by a vengeful ex with a PhD in cybersecurity.

AndrejMitrovic replied:
And here I thought she only cooked good pasta.

Geod24 commented on source/geod24/LocalRest.d:
```diff
@@ -698,17 +652,18 @@ public final class RemoteAPI (API, alias S = VibeJSONSerializer!()) : API
 
         Params:
-          tid = `geod24.concurrency.Tid` of the node.
+          listen_chn = `RestChannel` of the node.
```
Same remark here, connection seems more appropriate

Geod24 commented on source/geod24/LocalRest.d:
```diff
@@ -776,7 +734,10 @@ public final class RemoteAPI (API, alias S = VibeJSONSerializer!()) : API
             @trusted
         {
-            C.send(this.childTid, ShutdownCommand!API(callback, true));
+            if (isMainThread())
+                scheduler.start({ this.listen_chn.write(Variant(ShutdownCommand!API(callback, true))); });
+            else
+                this.listen_chn.write(Variant(ShutdownCommand!API(callback, true)));
```
I see many usages of this pattern... Why didn't you simply write a private method in the control interface which accepts a `Variant` and abstract this away ?

Geod24 commented on source/geod24/LocalRest.d:
```diff
@@ -935,56 +905,33 @@ public final class RemoteAPI (API, alias S = VibeJSONSerializer!()) : API
-
-                        // If the node already shut down, its MessageBox will be
-                        // closed. Detect it and notify the user.
-                        // Note that it might be expected that the remote died.
-                        if (!C.trySend(this.childTid, command))
-                            throw new Exception("Connection with peer closed");
+                        Response sendCommandAndWaitResponse ()
+                        {
+                            Variant var;
+                            if(!this.listen_chn.write(Variant(command)))
+                                throw new Exception("Connection with peer closed");
+                            if (resp_chn.read(var, this.timeout))
+                                return var.res;
+                            else
+                                return Response(Status.Timeout, SerializedData("Request timed-out"));
+                        }
```
Ah, here it is! That's the main issue: Sending a query to the node should not create a new channel. There should be two kinds of channels:
- 1) When `spawn`ing, you get a `BindChan` (or whatever name you see fit): this is the channel the node listens to for incoming connections.
- 2) When creating a client, you get a `ConnectionChan` (or `Connection`, or whatev) that correspond to the ephemeral connection between a client and a server (what you get after the server has `accept[4]()`ed the connection).

omerfirmak replied:
Well, my initial motivation was just to get a field test for channels. So I didn't want to re-design localrest around channels. I just replaced the obvious stuff. What you suggest is much better obviously :) I will give it a try.

omerfirmak replied:
@Geod24 , I gave this approach a shot today and I have some doubts about it. Having a TCP/HTTP like operation using Channels might be more trouble than it's worth. After giving it some thought, the benefits of this is kinda lost on me too :smile:  To me, it seems like only benefit is getting rid of the sender Tid in `Command`. We still have to route responses to whichever Fiber is waiting for it and we now have to manage 4 channels (ie. 2 `Connection`s, since both nodes in a pair can be the server) for **every node pair** in the system, instead of 1 `MessageBox` per node.

Geod24 replied:
> and we now have to manage 4 channels

4 channels ? Are you using channels as full-duplex connections or do you actually have a read and a write ?

Some aspects must be a bit tricky to manage, for example:
1) The `accept` channel cannot be full duplex, so a node actually need to send a channel or some kind of identifier.
2) Two fibers (client/server) can race on a single connection if used as full-duplex. This is alright as long as there is a way to know who sent/must receive the message, and they are able to put it back in the queue.

> instead of 1 `MessageBox` per node.

The main benefit is to be able to properly type the messages (the previous implementation completely bypasses the type system). There are other smaller benefit (easier to reason about, perhaps ability to control contention), but IMO type-safety is the main one.

Also I still don't see how even 4 channels / node brings overhead as compared to spawning a channel per query. I expect nodes will perform way more than 4 queries during their lifetime. Additionally, the overhead of a node is much higher than what the channel will bring.

omerfirmak replied:
> 4 channels ? Are you using channels as full-duplex connections or do you actually have a read and a write ?

For async communication, we would need 2 channels. One to send the request, second one to wait the response on. This makes a `Connection`. They get created when a new instance of `RemoteAPI` is created and `Connection` gets sent across the `accept` channel. This `Connection` is only good to send a request to the owner thread of the `RemoteAPI`. Every client should "initiate" a `Connection` to the server by instantiating a `RemoteAPI`, thus 2 `Connections` per every node pair. Using the same `Connection` for passing requests and responses in both directions would cause Nodes to race for initiating the `Connection` and require a mechanism to keep track of already existing `Connections`, but I am not sure how big of a problem that might be. This is almost a complete network stack :smile: 

> 4 channels / node brings

It is actually 4 channels / every possible node pair. 

Overhead of having persistent `Connection`s instead of short living Channels per query is in the implementation complexity. Node have to both manage incoming and outgoing `Connection`s separately. IMO, the biggest drawback of spawning a Channel per query is creating the OS Mutex and the GC allocation of the Channel itself, which I am not sure if either would be a significant bottleneck. But, since we know the lifetime of the response Channel we can allocate it on the stack. Also, OS Mutex can be replaced with a FiberMutex easily. **EDIT:** Maybe not so easily :P

Essentially what I am trying to say is that I would prefer simplicity over potential performance increase that we probably dont need right now. I might have gotten the whole `Connection` design wrong tho, it might be possible to implement it way simpler than I am trying to right now. :smile:

Geod24 replied:
> Using the same `Connection` for passing requests and responses in both directions would cause Nodes to race for initiating the `Connection` and require a mechanism to keep track of already existing `Connections`, but I am not sure how big of a problem that might be.

I'm not sure I follow. So let's say we have a network with 2 nodes, `A` and `B`. 
A` connects as a client to `B`, and `B` connects as a client to `A`.
`A` now have two `Connection`s: the client connection to `B`, and the server `Connection` to `B`. So does `B`.

When nothing happens, nodes listen to their server connection. When they get a query, they process it, and then write back on that same connection.
When doing a request as a client (e.g. during the processing of a query), they first write to the client connection. This should immediately block (yield to the event loop) until the query has been read. Then, they  read from that channel, which blocks (yield to the event loop) until someone writes.

The only moment nodes could ever race is when there's recursion.
E.g. `A [client]` -> `[server] B [client]` -> `[server] A [client]` -> `[server] B`.

However, you can match the  waiting fiber with the response if you associate them with an ID.

omerfirmak replied:
Moar commits for you to review @Geod24 @AndrejMitrovic  :)
The last commit should eventually replace [this](https://github.com/Geod24/localrest/pull/95/commits/91492cd1682efca7caec0fe8e9b72aa2960de6c9).

Geod24 commented on source/geod24/LocalRest.d:
```diff
@@ -2275,7 +2222,7 @@ unittest
     thread_joinAll();
 }
-
+/*
```
Sneaky, but no

Geod24 commented on source/geod24/Registry.d:
```diff
@@ -15,8 +15,8 @@ import geod24.concurrency;
 {
-    private Tid[string] tidByName;
-    private string[][Tid] namesByTid;
+    private Object[string] objectByName;
+    private string[][Object] namesByObject;
```
I believe, if you apply the suggestion to have a channel for the "listen" operation, you can have this typed (and same for the other `Object` types in `localrest`

Geod24 commented on source/geod24/concurrency.d:
```diff
@@ -1003,11 +1035,14 @@ protected:
+            if (this.shouldBlock())
+            {
+                this.registerToInfoFiber();
+                FiberScheduler.yieldAndThrow(new FiberBlockedException());
```
A trivial optimization would be to cache this instance, right ?

Geod24 commented on source/geod24/concurrency.d:
```diff
@@ -1097,59 +1133,124 @@ private:
+
+                // Fibers that block on a FiberBinarySemaphore throw an
+                // exception for scheduler to catch
+                if (auto block_ex = cast(FiberBlockedException) t)
```
If you cache the instance (assuming there's only one "in flight" at any time), you can also use `is` here, which does pointer comparison (as opposed to dynamic cast)

Geod24 commented on source/geod24/concurrency.d:
```diff
@@ -1501,6 +1501,176 @@ private:
+            sel_state = SelectState instace of the select call being executed
+            sel_id = id of the select call being executed
+
+    ***********************************************************************/
+
+    void selectWrite (void* ptr, SelectState sel_state, int sel_id);
+
+    /// Ditto
+    void selectRead (void* ptr, SelectState sel_state, int sel_id);
+}
+
+/***********************************************************************
+
+    An aggregate to hold neccessary information for a select operation
+
+***********************************************************************/
+
+public struct SelectEntry
+{
+    /// Reference to a Selectable object
+    Selectable selectable;
+
+    /// Pointer to the select data
+    void* select_data;
+
+    /***********************************************************************
+
+        Default ctor
+
+        Params:
+            selectable = A selectable interface reference
+            select_data = pointer to the data for the select operation
+
+    ***********************************************************************/
+
+    this (Selectable selectable, void* select_data)
+    {
+        this.selectable = selectable;
+        this.select_data = select_data;
+    }
```
D gives you a default ctor for struct. So I assume you only want it to be constructed with two parameter ?

AndrejMitrovic replied:
This was my thinking too. It prevents building a partially constructed aggregate. Kind of like `explicit` in C++.

omerfirmak replied:
Yes exactly, both parameters are required.

Geod24 reviewed (changes requested):
Amazing work! Sorry it took me so long to review. I focused on the commit that adapts the library, as it's the easiest to reason about. And as mentioned, there's a few issues in the approach, but nothing that can't be fixed :)

Geod24 commented:
Added the last commit to v0.x.x

omerfirmak committed:
Make Variant.Type camelCase and Variant.tag of type Type

omerfirmak committed:
Cache FiberBlockedException instance

omerfirmak committed:
Fix Agora tests failing with latest Agora

Because std.concurreny FiberScheduler lacks a schedule()
interface, tests emulate the behaviour with non-periodic
timer with 0.seconds timeout. 0.seconds timeout corresponds
to infinite timeout, so task never runs. sleep() can simply
return if called with 0.seconds duration.

omerfirmak committed:
Add a simple spin lock

omerfirmak committed:
Rename FiberBinarySemaphore to FiberBlocker

FiberBlocker will be the primitive that Fiber level
synchronization objects will build on.

omerfirmak committed:
Simplify isMainThread()

omerfirmak committed:
Report the invalid Variant.Type received

omerfirmak committed:
Eliminate some of the code duplications

omerfirmak commented:
~~I have just added the code for FiberMutex and FiberSemaphore but I am getting weird deadlocks in Agora tests. Do you guys see anything obviously wrong?
with OS locks, Fibers of same thread get to lock a Mutex at the same time, with FiberMutex they wont be able to do that. The bug might not be in the FiberMutex itself.~~

omerfirmak committed:
Add basic resource management to FiberScheduler

When a Fiber running under a FiberScheduler instance throws an
exception, scheduling loop is terminated. In that case all resources
held by the running Fibers should be released gracefully.

omerfirmak committed:
Implement FiberSemaphore/FiberMutex

omerfirmak committed:
Close maın channel before calling shutdown()

close() can block. When shutdown() is called and chn.close() blocks,
another task can be scheduled. Causing the shutdown()'ed node instance
to be used.

omerfirmak committed:
Add optional timeout to select()

omerfirmak committed:
Update LocalRest with Connections

Instead of response channels per Command, LocalRest now uses
persistent Connections between nodes.

AndrejMitrovic commented on source/geod24/concurrency.d:
```diff
@@ -963,6 +963,44 @@ class FiberScheduler
+        Remove Resource from the Resource list of runnning Fiber
+
+        Param:
+            r = Resource instace
```
What does it return?

AndrejMitrovic commented on source/geod24/concurrency.d:
```diff
@@ -1242,6 +1288,27 @@ private:
 
+    /***********************************************************************
+
+        Release all resources currently held by Fibers
```
by Fibers => by the provided running fiber?

AndrejMitrovic commented on source/geod24/concurrency.d:
```diff
@@ -2230,6 +2316,19 @@ final public class Channel (T) : Selectable
+        void release() nothrow
+        {
+            if (this.blocker.stopTimer())
+                this.pVal = cast(T*) 0x0; // Sanitize pVal so that we can catch illegal accesses
```
I think this can just be `= null;`?

AndrejMitrovic replied:
The cast will make the code non-`@safe`. It would be good to make most of the code we write `@safe`. But we can do it later too..

AndrejMitrovic commented on source/geod24/LocalRest.d:
```diff
@@ -148,13 +148,227 @@ private enum Status
+    void notifyWaiter (Response res)
+    {
+        // Is the waiting Fiber already blocked?
+        auto waiter = (res.id in this.waiting_list);
+        if (waiter !is null)
```
trick: 

```D
if (auto waiter = res.id in this.waiting_list)
{
}
else ...
```

AndrejMitrovic commented on source/geod24/LocalRest.d:
```diff
@@ -148,13 +148,227 @@ private enum Status
+
+    /*******************************************************************************
+
+        Close the `Channel`s associated with this `Connection`
```
It would be good to document what happens after the channel is closed. Do the other nodes's requests fail, or time out?

AndrejMitrovic commented on source/geod24/LocalRest.d:
```diff
@@ -443,10 +602,11 @@ public final class RemoteAPI (API, alias S = VibeJSONSerializer!()) : API
             cmd    = the command to run (contains the method name and the arguments)
             node   = the node to invoke the method on
             filter = used for filtering API calls (returns default response)
+            resp_chn = `Channel` to send the `Response` from
```
send to? or read from?

send from doesn't sound right.

AndrejMitrovic commented on source/geod24/LocalRest.d:
```diff
@@ -553,31 +711,78 @@ public final class RemoteAPI (API, alias S = VibeJSONSerializer!()) : API
             // Control the node behavior
             Control control;
 
+            struct AwaitingMessages
```
AwaitingMessages => AwaitingMessage

It's only a single message, and `await_msgs` is the array.

AndrejMitrovic commented on source/geod24/LocalRest.d:
```diff
@@ -553,31 +711,78 @@ public final class RemoteAPI (API, alias S = VibeJSONSerializer!()) : API
                     Variant msg;
-                    if (chn.read(msg, 10.msecs))
+
+                    C.SelectEntry[] read_list;
+                    C.SelectEntry[] write_list;
```
This will be quite GC-heavy with all the allocations. How about moving it oustide the while loop, resetting the lengths to zero and using `assumeSafeAppend`?

AndrejMitrovic commented on source/geod24/LocalRest.d:
```diff
@@ -553,31 +711,78 @@ public final class RemoteAPI (API, alias S = VibeJSONSerializer!()) : API
+                        continue;
+
+                    // Bind chn
+                    if (auto bind_chn = cast(BindChn) read_list[sel_ret.id].selectable)
```
`bid_chn` doesn't seem to be used here. So how about changing it to `if (cast(BindChn) read_list[sel_ret.id].selectable)`?

AndrejMitrovic commented on source/geod24/LocalRest.d:
```diff
@@ -553,31 +711,78 @@ public final class RemoteAPI (API, alias S = VibeJSONSerializer!()) : API
+                    // Command
+                    else if (auto comm_chn = cast(CommandChn) read_list[sel_ret.id].selectable)
                     {
+                        auto curr_conn = incoming_conns[comm_chn];
```
This can never be null?

omerfirmak replied:
Yep, that is non-null by design.

omerfirmak commented:
The weird thing is; with the last Connection changes, I am getting a 2x peak memory usage. The heap usage is barely increased. I think that the last changes are somehow holding the Fibers from getting GC'ed.

AndrejMitrovic commented:
> The weird thing is; with the last Connection changes, I am getting a 2x peak memory usage. The heap usage is barely increased. I think that the last changes are somehow holding the Fibers from getting GC'ed.

Do you know about the `-vgc` flag? It will list all allocations in the code.

AndrejMitrovic commented on source/geod24/LocalRest.d:
```diff
@@ -586,73 +706,110 @@ public final class RemoteAPI (API, alias S = VibeJSONSerializer!()) : API
             // Control the node behavior
             Control control;
 
+            struct AwaitingMessage
```
```suggestion
            static struct AwaitingMessage
```

D likes to add the context pointer even if it's unnecessary.

AndrejMitrovic commented on source/geod24/LocalRest.d:
```diff
@@ -1490,6 +1610,9 @@ unittest
 
         override void callFoo()
         {
+            if (!this.remote)
```
Why has this changed? It doesn't work when it's in the constructor?

omerfirmak replied:
it doesnt, `node_chn` gets initialized only after the first `spawn` finishes. Constructor runs before that.

AndrejMitrovic commented on source/geod24/concurrency.d:
```diff
@@ -926,84 +1038,138 @@ protected:
-            this.registerToInfoFiber();
-            this.limit = MonoTime.currTime + period;
-            FiberScheduler.yield();
+            if (period != Duration.init)
```
What about `Duration.max`?

AndrejMitrovic replied:
I mean, I think I saw code which handled both cases.

omerfirmak replied:
Only `Duration.init` is considered infinite timeout throughout localrest and concurrency.d. You must have seen it elsewhere?

Geod24 commented on source/geod24/LocalRest.d:
```diff
@@ -182,105 +148,268 @@ private enum Status
+/// Data sent by the caller
+private struct Command
 {
-    static if (T.length == 0)
-        size_t dummy;
-    T args;
+    /// Sequence id of Command
+    size_t id;
+    /// Method to call
+    string method;
+    /// Serialized arguments to the method
+    SerializedData args;
```
Can you move them back to where  they were so they don't show as diff ?

omerfirmak replied:
~~Once we squash the two LocalRest.d adaptation commits, they should be gone. I wanted to keep the two versions in the history.~~ Oh I see what you mean now :smile:

AndrejMitrovic commented on source/geod24/concurrency.d:
```diff
@@ -1012,56 +1178,154 @@ private:
         {
-            // Is Fiber waiting on a FiberBinarySemaphore?
-            if (auto sem = m_fibers[pos].sem)
+            while (!readyq.empty())
```
I'm a little bit confused about the nested loop. Can't you just remove it? The outer while loop handles everything.

omerfirmak replied:
We would need to replace the inner `while` with `if` and add an additional check to the `if` below to avoid sleeping while ready queue is not empty. I guess this could work

```diff
diff --git a/source/geod24/concurrency.d b/source/geod24/concurrency.d
index 947835d..909ef15 100644
--- a/source/geod24/concurrency.d
+++ b/source/geod24/concurrency.d
@@ -1190,7 +1190,7 @@ private:
 
         while (!this.readyq.empty() || !this.wait_list.empty())
         {
-            while (!readyq.empty())
+            if (!readyq.empty())
             {
                 InfoFiber cur_fiber = this.readyq.front();
                 this.readyq.removeFront();
@@ -1233,8 +1233,7 @@ private:
                 if (MonoTime.currTime >= earliest_timeout || this.sem.tryWait())
                     earliest_timeout = wakeFibers();
             }
-
-            if (!this.wait_list.empty())
+            else if (!this.wait_list.empty())
             {
                 Duration time_to_timeout = earliest_timeout - MonoTime.currTime;
 
@@ -1243,10 +1242,10 @@ private:
                     this.sem.wait();
                 else if (time_to_timeout > 0.seconds)
                     this.sem.wait(time_to_timeout);
-            }
 
-            // OS Thread woke up populate ready queue
-            earliest_timeout = wakeFibers();
+                // OS Thread woke up populate ready queue
+                earliest_timeout = wakeFibers();
+            }
         }
```

AndrejMitrovic commented on source/geod24/concurrency.d:
```diff
@@ -1012,56 +1178,154 @@ private:
+            if (!this.wait_list.empty())
             {
-                throw t;
+                Duration time_to_timeout = earliest_timeout - MonoTime.currTime;
```
Is it guaranteed that `earliest_timeout` will not equal `Duration.init` here? Actually I'm not even sure what happens when we subtract against that.

omerfirmak replied:
Yes, `earliest_timeout` is always > 0

Geod24 commented on source/geod24/concurrency.d:
```diff
@@ -1446,3 +1710,1545 @@ private:
+    }
+
+private:
+    /// Internal data storage
+    FixedRingBuffer!T buffer;
```
Do we really need to bring in a 315 LoC file just for this ? `FixedRingBuffer` seems rather generic, but we only iterate a single item at once and never resize the buffer. I only see 4 methods being used, `putBack`, `popFront`, `length`, `front`, and the initial capacity set.

AndrejMitrovic commented on source/geod24/concurrency.d:
```diff
@@ -1012,56 +1178,154 @@ private:
-                    pos = 0;
+                this.wait_list.popFirstOf(wait_range);
+                fiber.blocker = null;
+                this.readyq.insert(fiber);
```
What about popping from the `wait_range`? It's not needed here?

omerfirmak replied:
`popFirstOf` does that too. It removes both from the list and the range.

omerfirmak commented:
Did a rebase. Good news, we have merged more than half of the code yesterday.

omerfirmak committed:
Remove unused ThreadInfo from InfoFiber

omerfirmak committed:
Make FiberScheduler reuse terminated Fibers

Creating and destroying Fibers are costly due to the stack space
they allocate. We should try to reuse them as much as possible.

omerfirmak committed:
Break early from select loop when SelectState is already consumed

omerfirmak committed:
Dont create a queue entry for every FiberSemaphore.wait()

Creating a SemaphoreQueueEntry for every wait call, even if it
wont block, is REALLY heap heavy. We will only create entries for
blocking calls and let the user handle the cases where it would throw
an exception while holding a semaphore.

omerfirmak committed:
Make ChannelQueueEntry static

Saves 32 bytes per instance :shocked:

AndrejMitrovic commented:
> Did a rebase. Good news, we have merged more than half of the code yesterday.

Do you want to continue splitting this up into multiple PRs? I don't think I can review 20 commits in one go.
